narwhal/primary/src/core.rs

```diff
@@ -180,6 +179,51 @@ impl Core {
 
         self.highest_received_round = last_round_number;
         self.highest_processed_round = last_round_number;
+
+        // Get latest header from all peers and process it.
+        let peers: Vec<NetworkPublicKey> = self
+            .committee
+            .others_primaries(&self.name)
+            .into_iter()
+            .map(|(_, _, network_key)| network_key)
+            .collect();
+
+        let network = P2pNetwork::new(self.network.network());
+        let mut header_futures = FuturesUnordered::new();
+        let request = LatestHeaderRequest {};
+        for peer in peers.iter() {
+            let network = network.network();
+            let request = request.clone();
+
+            header_futures.push(async move {
+                let _ = &network;
+                network.get_latest_header(peer, request).await
+            });
+        }
+        while let Some(res) = header_futures.next().await {
+            match res {
+                Ok(response) => {
+                    if let Some(header) = response.header {
+                        let result = self.process_header(&header).await;
+                        if result.is_err() {
+                            error!(
+                                "error on recovery path processing latest header: {:?}",
+                                result.err()
+                            );
+                        }
+                    } else {
+                        trace!("peer's latest header was None on recovery path")
```

Comment:
maybe add in the peer ID to the error above & trace message here?

Reply:
Agreed. Also could you please switch that to a `debug` ? I think it's ok as we don't expect to be something noisy and might be useful to have in DevNet etc